source/common/upstream/upstream_impl.cc

```diff
@@ -919,9 +919,19 @@ ClusterImplBase::ClusterImplBase(
 
   auto socket_matcher = std::make_unique<TransportSocketMatcherImpl>(
       cluster.transport_socket_matches(), factory_context, socket_factory, *stats_scope);
-  info_ = std::make_unique<ClusterInfoImpl>(cluster, factory_context.clusterManager().bindConfig(),
-                                            runtime, std::move(socket_matcher),
-                                            std::move(stats_scope), added_via_api, factory_context);
+  auto& dispatcher = factory_context.dispatcher();
+  info_ = std::shared_ptr<const ClusterInfoImpl>(
```

Comment:
Moving the destructors for random objects like this one to the already CPU constrained main thread is likely to cause scale issues in Envoy.  I wonder if a more targetted fix in ~GrpcSubscriptionImpl that forces the unregistration to happen in the main thread would provide better performance while fully addressing the invalid cross thread interactions.

Reply:
I agree if the only goal is to fix the cross-thread event-socket-dispatcher binding.

But I also want to address the data race described here. https://github.com/envoyproxy/envoy/issues/13209

Combining these two, I think it's better to follow the pattern that config~ish patten objects be destroyed at the master thread.

We can resolve the master thread task overhead in the future. The majority of the cluster infos are indeeded destroyed on master thread anyway so that the bug was not noticed.

Reply:
I missed that ClusterInfoImpl is created in the main thread and info_ holds a shared_ptr, not an unique_ptr.

I agree that it is unsafe to do this deletion outside the main thread, so I guess that this custom destructor to move the operation to the main thread is unavoidable.